src/v/coding-style.md

```diff
@@ -6,12 +6,50 @@ The document covers style and convention not enforced by clang-format.
 
 Header files have the `.h` extension, source files use the `.cc` extension.
 
+Public header files should go in an `includes` directory within the module they belong to. 
+For example, a module named `foo` should have public header files within `src/v/foo/include/foo`.
+
 Test files go in the `tests` directory of the module they belong to. That is,
 the tests for a feature named `foo` and located in `src/v/foo` will be located
 in `src/v/foo/tests`.
 
 Use `snake_case` naming convention by default for files and language identifiers.
 
+## Documentation
+
+### Modules
+
+Modules themselves should have a `README.md` file explaining the purpose of the module,
+it's API surface, and an architectural overview of that module (mentioning major classes),
+etc. This will aid people new to the module in understanding the lay of the land. When in doubt
+of if information should be placed in a class or in a README, prefer moving the documentation
+as close to the source as possible (inside the class documentation). We want to prioritize
+documentation not becoming stale - and also not cognitively overload the reader of the module
+README.
+
+### Classes
+
+Classes in public headers should be documented, plain structs that only hold data are not
+required to be documented, but it is often helpful to add. Writing points when documenting
+a given class:
+
+* Lifecycle of the class
+* Place within the larger module
+* How to use it correctly
+* Broad strokes of implmentation details
+* Which cores it exists on (if a sharded service)
+* If it is a cluster/node/shard view of information
+
+While only required in public headers, it's also encouraged to add this information to key classes 
```

Comment:
Personally I think we should document _all_ classes. Ultimately I feel the "module boundary" is not a very important one here, inside Redpanda, for the purposes of documentation because in general core is working across modules, i.e., it is not as if the public header documentation is consumed by a different set of people. The same reasoning for documenting "public" headers mostly applies to classes anywhere else?